sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,455 @@
+        {
+            request.Content.WriteTo(memoryStream, cancellationToken);
+        }
+        var bytes = memoryStream.ToArray();
+
+        // Try to extract a text encoding from the headers
+        Encoding? requestTextEncoding = null;
+
+        if (request.Headers.TryGetValue("Content-Type", out var contentType) && contentType != null)
+        {
+            ContentTypeUtilities.TryGetTextEncoding(contentType, out requestTextEncoding);
+        }
+
+        // Log to event source
+        Singleton.RequestContent(requestId ?? string.Empty, bytes, requestTextEncoding);
+    }
+
+    private void LogResponse(PipelineResponse response, string? responseId, double elapsed)
+    {
+        bool isError = response.IsError;
+        if (isError)
+        {
+            Singleton.ErrorResponse(response, responseId ?? string.Empty, Sanitizer!, elapsed);
+        }
+        else
+        {
+            Singleton.Response(response, responseId ?? string.Empty, Sanitizer!, elapsed);
+        }
+    }
+
+    private async Task LogResponseContent(PipelineResponse response, string? responseId, bool bufferResponse, bool async, CancellationToken cancellationToken)
+    {
+        // If logging content is disabled, or the request content is null, or informational logs are not enabled, return.
+        // Checking the log level prevents expensive operations from being executed
+        // TODO - consider moving this down to LoggingStream
+        var logErrorResponseContent = response.IsError && Singleton.IsEnabled(EventLevel.Warning, EventKeywords.All);
+        var logResponseContent = Singleton.IsEnabled(EventLevel.Informational, EventKeywords.All) || logErrorResponseContent;
+
+        if (!_logContent || !logResponseContent)
+        {
+            return; // nothing to log
+        }
+
+        // Determine if the content stream should be wrapped in a logging stream, which logs the stream in blocks as it is being read
+        var wrapContent =
+            !bufferResponse // Content is not buffered - content should be in the content stream
+            && response.ContentStream != null // Content stream is available
+            && response.ContentStream?.CanSeek == false; // Content stream is not seekable - if it's seekable we can just log directly
+
+        // Try to extract a text encoding from the headers
+        Encoding? responseTextEncoding = null;
+
+        if (response.Headers.TryGetValue("Content-Type", out var contentType) && contentType != null)
+        {
+            ContentTypeUtilities.TryGetTextEncoding(contentType, out responseTextEncoding);
+        }
+
+        if (wrapContent)
+        {
+            response.ContentStream = new LoggingStream(responseId ?? string.Empty, _maxLength, response.ContentStream!, response.IsError, responseTextEncoding);
+        }
+        else
+        {
+            byte[]? bytes = null;
+            if (bufferResponse)
+            {
+                // TODO - is this the best way to do this
```

Comment:
Would something like the following work?

```csharp
bytes = response.Content.ToMemory().Span.Slice(0, _maxLength);
```

or
 
```csharp
bytes = ((ReadOnlySpan<byte>)response.Content).Slice(0, _maxLength).ToArray();
```

?

ToArray will allocate the bytes, but that has to happen somewhere once.

Reply:
Updated!